src/main/kotlin/io/github/yuutoproject/yuutobot/commands/Help.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Open source bot built by and for the Camp Buddy Discord Fan Server.
+ *     Copyright (C) 2020  Kyuuto-devs
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package io.github.yuutoproject.yuutobot.commands
+
+import io.github.yuutoproject.yuutobot.CommandManager
+import io.github.yuutoproject.yuutobot.commands.base.AbstractCommand
+import io.github.yuutoproject.yuutobot.commands.base.CommandCategory
+import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
+
+class Help(private val commandManager: CommandManager) : AbstractCommand(
+    "help",
+    CommandCategory.INFO,
+    "Get the usage of any command",
+    "[command]",
+    "Run `help list` to list possible commands"
+) {
+    override val aliases = arrayOf("usage", "commands")
+
+    override fun run(args: MutableList<String>, event: GuildMessageReceivedEvent) {
+        val commands = this.commandManager.commands
+        val aliases = this.commandManager.aliases
+
+        val commandName = args.getOrElse(0) { "list" }
+
+        if (commandName == "list") {
+            val message = event.channel.sendMessage(
+                "Here is a list of all commands and their descriptions:\n"
+            )
+
+            val groups = commands.values.groupBy(AbstractCommand::category)
+
+            for ((category, commandsInCategory) in groups) {
+                message.append("\n$category\n")
+
+                for (command in commandsInCategory) {
+                    message.append("`${command.name}` - ${command.description}\n")
+                }
+            }
+
+            message.queue()
+            return
+        }
+
+        // No need to worry about indexing with null, it just returns null
+        val command = commands[commandName] ?: commands[aliases[commandName]] ?: run {
```

Comment:
I would make a method in the command manager that fetches the command for you or returns null, that way you have less duped code.
Also, does this work when the command is unknown? I haven't seen you test this

Reply:
Yes, if `commandName` is not a registered command or alias, then the final `run` statement will run. I think Kotlin is smart enough to skip evaluating `commands[aliases[commandName]]` instead of throwing an error because of `commands[null]`.

Reply:
>  I think Kotlin is smart enough to skip evaluating commands[aliases[commandName]] instead of throwing an error because of commands[null]

You can check what it compiles to by going to `tools > kotlin > view kotlin bytecode` and then hitting decompile in the tab that opens